kclvm/tools/src/LSP/src/document_symbol.rs

```diff
@@ -0,0 +1,96 @@
+use kclvm_sema::resolver::scope::ScopeObject;
+use kclvm_sema::resolver::scope::ScopeObjectKind;
+use lsp_types::Range;
+use lsp_types::{DocumentSymbol, DocumentSymbolResponse, SymbolKind};
+
+use crate::to_lsp::lsp_pos;
+
+pub(crate) fn document_symbol(
+    file: &str,
+    _program: &Program,
+    prog_scope: &ProgramScope,
+) -> Option<lsp_types::DocumentSymbolResponse> {
+    let mut documentsymbols: Vec<DocumentSymbol> = vec![];
+    let scope = prog_scope.scope_map.get("__main__").unwrap().borrow();
+    // Get variable in scope
+    for obj in scope.elems.values().filter(|obj| {
+        obj.borrow().start.filename == file
+            && obj.borrow().kind == kclvm_sema::resolver::scope::ScopeObjectKind::Variable
+    }) {
+        documentsymbols.push(scope_obj_to_document_symbol(obj.borrow().clone()).unwrap())
+    }
+    // Get schema definition in scope
+    for child in scope
+        .children
+        .iter()
+        .filter(|child| child.borrow().start.filename == file)
+    {
+        if let Some(symbol) = schema_scope_to_document_symbol(child.borrow().clone()) {
+            documentsymbols.push(symbol)
+        }
+    }
+    Some(DocumentSymbolResponse::Nested(documentsymbols))
+}
+
+#[allow(deprecated)]
+fn schema_scope_to_document_symbol(scope: Scope) -> Option<DocumentSymbol> {
+    if let ScopeKind::Schema(schema_name) = &scope.kind {
+        let range = Range {
+            start: lsp_pos(&scope.start),
+            end: lsp_pos(&scope.end),
+        };
+        Some(DocumentSymbol {
+            name: schema_name.clone(),
+            kind: SymbolKind::STRUCT,
+            range,
+            selection_range: range,
+            children: Some(
+                scope
+                    .elems
+                    .iter()
+                    .map(|(_, obj)| scope_obj_to_document_symbol(obj.borrow().clone()))
+                    .filter_map(|symbol| symbol)
+                    .collect(),
+            ),
+            detail: None,
+            tags: None,
+            deprecated: None,
+        })
+    } else {
+        None
+    }
+}
+
+#[allow(deprecated)]
+fn scope_obj_to_document_symbol(obj: ScopeObject) -> Option<DocumentSymbol> {
+    if let Some(kind) = scope_obj_kind_to_document_symbol_kind(obj.kind) {
+        let range = Range {
+            start: lsp_pos(&obj.start),
+            end: lsp_pos(&obj.end),
+        };
+        Some(DocumentSymbol {
+            name: obj.name.clone(),
+            kind,
+            range,
+            selection_range: range,
+            detail: None,
+            tags: None,
+            children: None,
+            deprecated: None,
+        })
+    } else {
+        None
+    }
+}
+
+fn scope_obj_kind_to_document_symbol_kind(kind: ScopeObjectKind) -> Option<SymbolKind> {
+    match kind {
+        ScopeObjectKind::Variable => Some(SymbolKind::VARIABLE),
+        ScopeObjectKind::Attribute => Some(SymbolKind::PROPERTY),
+        _ => None,
```

Comment:
Why is `SymbolKind::STRUCT` not handled here?

Reply:
We use ScopeObjectKind::Definition to represent Struct/Schema, but in this structure, there is no elem/child to represent the schema attribute and their location information. So generate SymbolKind::Struct from ScopeKind::Schema instead of ScopeObjectKind::Definition